File: autozig-app/src/zig/plugin_group.zig
const std = @import("std");
const ZigApp = @import("app.zig").ZigApp;
const ZigPlugin = @import("plugin.zig").ZigPlugin;

/// Plugin entry in a group - tracks plugin and enabled state
pub const PluginEntry = struct {
    plugin: *ZigPlugin,
    enabled: bool,
    type_id: u64,
};

/// Plugin Group Builder - manages ordered collection of plugins
pub const PluginGroupBuilder = struct {
    allocator: std.mem.Allocator,
    group_name: []const u8,
    plugins: std.AutoHashMap(u64, PluginEntry),
    order: std.ArrayList(u64),

    pub fn create(allocator: std.mem.Allocator, name: []const u8) !*PluginGroupBuilder {
        const builder = try allocator.create(PluginGroupBuilder);

        const name_copy = try allocator.dupe(u8, name);

        builder.* = PluginGroupBuilder{
            .allocator = allocator,
            .group_name = name_copy,
            .plugins = std.AutoHashMap(u64, PluginEntry).init(allocator),
            .order = std.ArrayList(u64){},
        };

        return builder;
    }

    pub fn destroy(self: *PluginGroupBuilder) void {
        self.allocator.free(self.group_name);

        // Note: plugins are owned by the App, we only store references
        self.plugins.deinit();
        self.order.deinit(self.allocator);

        self.allocator.destroy(self);
    }

    /// Check if builder contains a plugin by type_id
    pub fn contains(self: *PluginGroupBuilder, type_id: u64) bool {
        return self.plugins.contains(type_id);
    }

    /// Check if a plugin is enabled
    pub fn isEnabled(self: *PluginGroupBuilder, type_id: u64) bool {
        if (self.plugins.get(type_id)) |entry| {
            return entry.enabled;
        }
        return false;
    }

    /// Find index of a plugin in the order list
    fn indexOf(self: *PluginGroupBuilder, type_id: u64) ?usize {
        for (self.order.items, 0..) |id, i| {
            if (id == type_id) {
                return i;
            }
        }
        return null;
    }

    /// Add plugin at the end of the builder
    pub fn add(self: *PluginGroupBuilder, plugin: *ZigPlugin, type_id: u64) !void {
        // Remove from previous position if it exists
        if (self.indexOf(type_id)) |old_index| {
            _ = self.order.orderedRemove(old_index);
        }

        // Add to end
        try self.order.append(self.allocator, type_id);

        const entry = PluginEntry{
            .plugin = plugin,
            .enabled = true,
            .type_id = type_id,
        };

        try self.plugins.put(type_id, entry);
    }

    /// Add plugin before target plugin
    pub fn addBefore(
        self: *PluginGroupBuilder,
        plugin: *ZigPlugin,
        type_id: u64,
        target_type_id: u64,
    ) !bool {
        const target_index = self.indexOf(target_type_id) orelse return false;

        // Remove from previous position if it exists
        if (self.indexOf(type_id)) |old_index| {
            _ = self.order.orderedRemove(old_index);
        }

        // Insert before target
        try self.order.insert(self.allocator, target_index, type_id);

        const entry = PluginEntry{
            .plugin = plugin,
            .enabled = true,
            .type_id = type_id,
        };

        try self.plugins.put(type_id, entry);
        return true;
    }

    /// Add plugin after target plugin
    pub fn addAfter(
        self: *PluginGroupBuilder,
        plugin: *ZigPlugin,
        type_id: u64,
        target_type_id: u64,
    ) !bool {
        const target_index = self.indexOf(target_type_id) orelse return false;

        // Remove from previous position if it exists
        if (self.indexOf(type_id)) |old_index| {
            _ = self.order.orderedRemove(old_index);
        }

        // Insert after target
        const insert_index = target_index + 1;
        try self.order.insert(self.allocator, insert_index, type_id);

        const entry = PluginEntry{
            .plugin = plugin,
            .enabled = true,
            .type_id = type_id,
        };

        try self.plugins.put(type_id, entry);
        return true;
    }

    /// Enable a plugin
    pub fn enable(self: *PluginGroupBuilder, type_id: u64) !void {
        if (self.plugins.getPtr(type_id)) |entry| {
            entry.enabled = true;
        } else {
            return error.PluginNotFound;
        }
    }

    /// Disable a plugin
    pub fn disable(self: *PluginGroupBuilder, type_id: u64) !void {
        if (self.plugins.getPtr(type_id)) |entry| {
            entry.enabled = false;
        } else {
            return error.PluginNotFound;
        }
    }

    /// Set/replace plugin value
    pub fn set(self: *PluginGroupBuilder, plugin: *ZigPlugin, type_id: u64) !bool {
        if (self.plugins.getPtr(type_id)) |entry| {
            entry.plugin = plugin;
            return true;
        }
        return false;
    }

    /// Finish building and add all enabled plugins to app
    pub fn finish(self: *PluginGroupBuilder, app: *ZigApp) !void {
        for (self.order.items) |type_id| {
            if (self.plugins.get(type_id)) |entry| {
                if (entry.enabled) {
                    _ = try app.addPlugin(entry.plugin);
                }
            }
        }
    }

    /// Get number of plugins in builder
    pub fn len(self: *PluginGroupBuilder) usize {
        return self.order.items.len;
    }

    /// Get number of enabled plugins
    pub fn enabledCount(self: *PluginGroupBuilder) usize {
        var count: usize = 0;
        for (self.order.items) |type_id| {
            if (self.plugins.get(type_id)) |entry| {
                if (entry.enabled) {
                    count += 1;
                }
            }
        }
        return count;
    }
};

// FFI exports
export fn plugin_group_builder_create(name_ptr: [*]const u8, name_len: usize) ?*PluginGroupBuilder {
    const allocator = std.heap.page_allocator;
    const name = name_ptr[0..name_len];
    return PluginGroupBuilder.create(allocator, name) catch null;
}

export fn plugin_group_builder_destroy(builder: *PluginGroupBuilder) void {
    builder.destroy();
}

export fn plugin_group_builder_contains(builder: *PluginGroupBuilder, type_id: u64) bool {
    return builder.contains(type_id);
}

export fn plugin_group_builder_is_enabled(builder: *PluginGroupBuilder, type_id: u64) bool {
    return builder.isEnabled(type_id);
}

export fn plugin_group_builder_add(
    builder: *PluginGroupBuilder,
    plugin: *ZigPlugin,
    type_id: u64,
) bool {
    builder.add(plugin, type_id) catch return false;
    return true;
}

export fn plugin_group_builder_add_before(
    builder: *PluginGroupBuilder,
    plugin: *ZigPlugin,
    type_id: u64,
    target_type_id: u64,
) bool {
    return builder.addBefore(plugin, type_id, target_type_id) catch false;
}

export fn plugin_group_builder_add_after(
    builder: *PluginGroupBuilder,
    plugin: *ZigPlugin,
    type_id: u64,
    target_type_id: u64,
) bool {
    return builder.addAfter(plugin, type_id, target_type_id) catch false;
}

export fn plugin_group_builder_enable(builder: *PluginGroupBuilder, type_id: u64) bool {
    builder.enable(type_id) catch return false;
    return true;
}

export fn plugin_group_builder_disable(builder: *PluginGroupBuilder, type_id: u64) bool {
    builder.disable(type_id) catch return false;
    return true;
}

export fn plugin_group_builder_set(
    builder: *PluginGroupBuilder,
    plugin: *ZigPlugin,
    type_id: u64,
) bool {
    return builder.set(plugin, type_id) catch false;
}

export fn plugin_group_builder_finish(builder: *PluginGroupBuilder, app: *ZigApp) bool {
    builder.finish(app) catch return false;
    return true;
}

export fn plugin_group_builder_len(builder: *PluginGroupBuilder) usize {
    return builder.len();
}

export fn plugin_group_builder_enabled_count(builder: *PluginGroupBuilder) usize {
    return builder.enabledCount();
}

File: autozig-app/src/plugin_group.rs
//! Plugin Group system for organizing and configuring application plugins
//!
//! Provides PluginGroup trait and PluginGroupBuilder for managing collections of plugins
//! with ordering constraints and enable/disable functionality.

#![forbid(unsafe_code)]

use crate::{App, Plugin, ZigPlugin};
use autozig::include_zig;
use core::any::TypeId;
use core::marker::PhantomData;
use core::ptr::NonNull;

/// Opaque Zig PluginGroupBuilder type
#[repr(C)]
pub struct ZigPluginGroupBuilder {
    _private: [u8; 0],
}

// Include Zig FFI functions
include_zig!("src/zig/plugin_group.zig", {
    fn plugin_group_builder_create(name_ptr: *const u8, name_len: usize) -> *mut ZigPluginGroupBuilder;
    fn plugin_group_builder_destroy(builder: *mut ZigPluginGroupBuilder);
    fn plugin_group_builder_contains(builder: *mut ZigPluginGroupBuilder, type_id: u64) -> bool;
    fn plugin_group_builder_is_enabled(builder: *mut ZigPluginGroupBuilder, type_id: u64) -> bool;
    fn plugin_group_builder_add(builder: *mut ZigPluginGroupBuilder, plugin: *mut ZigPlugin, type_id: u64) -> bool;
    fn plugin_group_builder_add_before(builder: *mut ZigPluginGroupBuilder, plugin: *mut ZigPlugin, type_id: u64, target_type_id: u64) -> bool;
    fn plugin_group_builder_add_after(builder: *mut ZigPluginGroupBuilder, plugin: *mut ZigPlugin, type_id: u64, target_type_id: u64) -> bool;
    fn plugin_group_builder_enable(builder: *mut ZigPluginGroupBuilder, type_id: u64) -> bool;
    fn plugin_group_builder_disable(builder: *mut ZigPluginGroupBuilder, type_id: u64) -> bool;
    fn plugin_group_builder_set(builder: *mut ZigPluginGroupBuilder, plugin: *mut ZigPlugin, type_id: u64) -> bool;
    fn plugin_group_builder_finish(builder: *mut ZigPluginGroupBuilder, app: *mut crate::ZigApp) -> bool;
    fn plugin_group_builder_len(builder: *mut ZigPluginGroupBuilder) -> usize;
    fn plugin_group_builder_enabled_count(builder: *mut ZigPluginGroupBuilder) -> usize;
});

/// Combines multiple [`Plugin`]s into a single unit
///
/// This trait allows organizing related plugins together and provides
/// methods for configuring them before adding to an App.
pub trait PluginGroup: Sized {
    /// Configures the [`Plugin`]s that are to be added
    fn build(self) -> PluginGroupBuilder;
    
    /// Configures a name for the [`PluginGroup`] which is primarily used for debugging
    fn name() -> &'static str {
        core::any::type_name::<Self>()
    }
    
    /// Sets the value of the given [`Plugin`], if it exists
    fn set<T: Plugin>(self, plugin: T) -> PluginGroupBuilder {
        self.build().set(plugin)
    }
}

/// Facilitates the creation and configuration of a [`PluginGroup`]
///
/// Provides build ordering to ensure that [`Plugin`]s which produce/require resources
/// are built before/after dependent/depending [`Plugin`]s. [`Plugin`]s inside the group
/// can be disabled, enabled or reordered.
pub struct PluginGroupBuilder {
    inner: NonNull<ZigPluginGroupBuilder>,
}

impl PluginGroupBuilder {
    /// Start a new builder for the [`PluginGroup`]
    pub fn start<PG: PluginGroup>() -> Self {
        let name = PG::name();
        let ptr = plugin_group_builder_create(name.as_ptr(), name.len());
        Self {
            inner: NonNull::new(ptr).expect("plugin group builder creation failed"),
        }
    }
    
    /// Checks if the [`PluginGroupBuilder`] contains the given [`Plugin`]
    pub fn contains<T: Plugin>(&self) -> bool {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        plugin_group_builder_contains(self.inner.as_ptr(), type_id)
    }
    
    /// Returns `true` if the [`PluginGroupBuilder`] contains the given [`Plugin`] and it's enabled
    pub fn enabled<T: Plugin>(&self) -> bool {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        plugin_group_builder_is_enabled(self.inner.as_ptr(), type_id)
    }
    
    /// Adds the plugin [`Plugin`] at the end of this [`PluginGroupBuilder`]
    ///
    /// If the plugin was already in the group, it is removed from its previous place.
    pub fn add<T: Plugin>(self, plugin: T) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        let plugin_ptr = plugin.into_zig_plugin();
        
        plugin_group_builder_add(self.inner.as_ptr(), plugin_ptr, type_id);
        self
    }
    
    /// Adds a [`Plugin`] in this [`PluginGroupBuilder`] before the plugin of type `Target`
    ///
    /// If the plugin was already in the group, it is removed from its previous place.
    ///
    /// # Panics
    ///
    /// Panics if `Target` is not already in this [`PluginGroupBuilder`].
    pub fn add_before<Target: Plugin, T: Plugin>(self, plugin: T) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        let target_type_id = type_id_to_u64(TypeId::of::<Target>());
        let plugin_ptr = plugin.into_zig_plugin();
        
        if !plugin_group_builder_add_before(
            self.inner.as_ptr(),
            plugin_ptr,
            type_id,
            target_type_id,
        ) {
            panic!(
                "Plugin does not exist in group: {}",
                core::any::type_name::<Target>()
            );
        }
        
        self
    }
    
    /// Adds a [`Plugin`] in this [`PluginGroupBuilder`] after the plugin of type `Target`
    ///
    /// If the plugin was already in the group, it is removed from its previous place.
    ///
    /// # Panics
    ///
    /// Panics if `Target` is not already in this [`PluginGroupBuilder`].
    pub fn add_after<Target: Plugin, T: Plugin>(self, plugin: T) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        let target_type_id = type_id_to_u64(TypeId::of::<Target>());
        let plugin_ptr = plugin.into_zig_plugin();
        
        if !plugin_group_builder_add_after(
            self.inner.as_ptr(),
            plugin_ptr,
            type_id,
            target_type_id,
        ) {
            panic!(
                "Plugin does not exist in group: {}",
                core::any::type_name::<Target>()
            );
        }
        
        self
    }
    
    /// Enables a [`Plugin`]
    ///
    /// [`Plugin`]s within a [`PluginGroup`] are enabled by default. This function is used to
    /// opt back in to a [`Plugin`] after disabling it.
    ///
    /// # Panics
    ///
    /// Panics if there are no plugins of type `T` in this group.
    pub fn enable<T: Plugin>(self) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        
        if !plugin_group_builder_enable(self.inner.as_ptr(), type_id) {
            panic!(
                "Cannot enable a plugin that does not exist: {}",
                core::any::type_name::<T>()
            );
        }
        
        self
    }
    
    /// Disables a [`Plugin`], preventing it from being added to the [`App`]
    ///
    /// The disabled [`Plugin`] keeps its place in the [`PluginGroup`], so it can
    /// still be used for ordering with [`add_before`](Self::add_before) or
    /// [`add_after`](Self::add_after), or it can be re-enabled.
    ///
    /// # Panics
    ///
    /// Panics if there are no plugins of type `T` in this group.
    pub fn disable<T: Plugin>(self) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        
        if !plugin_group_builder_disable(self.inner.as_ptr(), type_id) {
            panic!(
                "Cannot disable a plugin that does not exist: {}",
                core::any::type_name::<T>()
            );
        }
        
        self
    }
    
    /// Sets the value of the given [`Plugin`], if it exists
    ///
    /// # Panics
    ///
    /// Panics if the [`Plugin`] does not exist.
    pub fn set<T: Plugin>(self, plugin: T) -> Self {
        let type_id = type_id_to_u64(TypeId::of::<T>());
        let plugin_ptr = plugin.into_zig_plugin();
        
        if !plugin_group_builder_set(self.inner.as_ptr(), plugin_ptr, type_id) {
            panic!(
                "{} does not exist in this PluginGroup",
                core::any::type_name::<T>()
            );
        }
        
        self
    }
    
    /// Consumes the [`PluginGroupBuilder`] and builds the contained [`Plugin`]s
    /// in the order specified
    ///
    /// # Panics
    ///
    /// Panics if one of the plugins in the group was already added to the application.
    pub fn finish(self, app: &mut App) {
        if !plugin_group_builder_finish(self.inner.as_ptr(), app.inner.as_ptr()) {
            panic!("Failed to finish plugin group");
        }
        // Builder is consumed, drop will clean up
    }
    
    /// Get number of plugins in the builder
    pub fn len(&self) -> usize {
        plugin_group_builder_len(self.inner.as_ptr())
    }
    
    /// Check if the builder is empty
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    
    /// Get number of enabled plugins
    pub fn enabled_count(&self) -> usize {
        plugin_group_builder_enabled_count(self.inner.as_ptr())
    }
}

impl Drop for PluginGroupBuilder {
    fn drop(&mut self) {
        plugin_group_builder_destroy(self.inner.as_ptr());
    }
}

// Implement PluginGroup for PluginGroupBuilder to allow chaining
impl PluginGroup for PluginGroupBuilder {
    fn build(self) -> PluginGroupBuilder {
        self
    }
}

/// Helper function to convert TypeId to u64
fn type_id_to_u64(type_id: TypeId) -> u64 {
    use core::hash::Hasher;
    
    let mut hasher = TypeIdHasher::default();
    core::hash::Hash::hash(&type_id, &mut hasher);
    hasher.finish()
}

/// Simple TypeId hasher implementation (completely safe)
#[derive(Default)]
struct TypeIdHasher {
    state: u64,
}

impl core::hash::Hasher for TypeIdHasher {
    fn write(&mut self, bytes: &[u8]) {
        // FNV-1a hash algorithm (completely safe implementation)
        const FNV_OFFSET_BASIS: u64 = 0xcbf29ce484222325;
        const FNV_PRIME: u64 = 0x100000001b3;
        
        if self.state == 0 {
            self.state = FNV_OFFSET_BASIS;
        }
        
        for &byte in bytes {
            self.state ^= byte as u64;
            self.state = self.state.wrapping_mul(FNV_PRIME);
        }
    }
    
    fn finish(&self) -> u64 {
        self.state
    }
}

/// Extension trait for App to support plugin groups
pub trait PluginGroupExt {
    /// Add a [`PluginGroup`] to the application
    fn add_plugin_group<G: PluginGroup>(&mut self, group: G) -> &mut Self;
}

impl PluginGroupExt for App {
    fn add_plugin_group<G: PluginGroup>(&mut self, group: G) -> &mut Self {
        let builder = group.build();
        builder.finish(self);
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::SimplePlugin;
    
    #[test]
    fn test_plugin_group_builder_create() {
        let builder = PluginGroupBuilder::start::<TestPluginGroup>();
        assert_eq!(builder.len(), 0);
        assert!(builder.is_empty());
    }
    
    #[test]
    fn test_plugin_group_add() {
        let plugin = SimplePlugin::new("test", |_| {});
        let builder = PluginGroupBuilder::start::<TestPluginGroup>().add(plugin);
        
        assert_eq!(builder.len(), 1);
        assert!(!builder.is_empty());
        assert!(builder.contains::<SimplePlugin>());
    }
    
    struct TestPluginGroup;
    
    impl PluginGroup for TestPluginGroup {
        fn build(self) -> PluginGroupBuilder {
            PluginGroupBuilder::start::<Self>()
        }
    }
}